dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-raft/src/main/java/org/apache/dolphinscheduler/plugin/registry/raft/RaftRegistry.java

```diff
@@ -0,0 +1,228 @@
+                .withRpcOptions(rpcOptions)
+                .config();
+        this.kvStore = new DefaultRheaKVStore();
+        this.kvStore.init(opts);
+        log.info("kvStore started...");
+        if (!properties.getModule().equalsIgnoreCase(API_TYPE)) {
+            this.subscribeListenerManager = new SubscribeListenerManager(properties, kvStore);
+            subscribeListenerManager.start();
+        }
+    }
+
+    @Override
+    public boolean subscribe(String path, SubscribeListener listener) {
+        return subscribeListenerManager.addSubscribeListener(path, listener);
+    }
+
+    @Override
+    public void unsubscribe(String path) {
+        subscribeListenerManager.removeSubscribeListener(path);
+    }
+
+    @Override
+    public void addConnectionStateListener(ConnectionListener listener) {
+        final String groupId = properties.getClusterName() + "--1";
+        final Node node = NodeManager.getInstance().get(groupId, new PeerId((properties.getServerAddress()), properties.getServerPort()));
+        node.addReplicatorStateListener(new RaftConnectionStateListener(listener));
+    }
+
+    @Override
+    public String get(String key) {
+        return readUtf8(kvStore.bGet(key));
+    }
+
+    @Override
+    public void put(String key, String value, boolean deleteOnDisconnect) {
+        kvStore.bPut(key, writeUtf8(value));
+        if (key.startsWith(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS + Constants.SINGLE_SLASH)) {
+            addWorkerGroup(key);
+        }
+    }
+
+    private void addWorkerGroup(String key) {
+        List<String> workerGroupList = getWorkerGroups();
+        String workerGroup = key.substring(key.indexOf(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS)
+                + Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS.length() + 1, key.lastIndexOf(Constants.SINGLE_SLASH));
+        if (!workerGroupList.contains(workerGroup)) {
+            workerGroupList.add(workerGroup);
+            kvStore.bPut(REGISTRY_DOLPHINSCHEDULER_WORKER_GROUPS, writeUtf8(JSONUtils.toJsonString(workerGroupList)));
+        }
+    }
+
+    private List<String> getWorkerGroups() {
+        final String storedWorkerGroup = readUtf8(kvStore.bGet(REGISTRY_DOLPHINSCHEDULER_WORKER_GROUPS));
+        if (StringUtils.isEmpty(storedWorkerGroup)) {
+            return new ArrayList<>();
+        }
+        return new ArrayList<>(JSONUtils.toList(storedWorkerGroup, String.class));
+    }
+
+    @Override
+    public void delete(String key) {
+        kvStore.bDelete(key);
+        final DistributedLock<byte[]> distributedLock = distributedLockMap.get(key);
+        if (distributedLock != null) {
+            distributedLock.unlock();
+        }
+        distributedLockMap.remove(key);
+    }
+
+    @Override
+    public Collection<String> children(String key) {
+        List<String> children = new ArrayList<>();
+        if (key.equals(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS)) {
```

Comment:
Why we need to do this judge here, this method is just want to get the children under given key.

Reply:
Because jraft not have getchildren function, so store workgroups separately